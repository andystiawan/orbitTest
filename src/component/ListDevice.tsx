import { useState } from 'react';
import {
  FlatList,
  Image,
  ListRenderItem,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { DHeight, DWidth } from './Dimension';
import { SortDevice } from './Sort';

const height: number = DHeight;
const width: number = DWidth;

const filter = [
  { value: '', label: 'Default   ' },
  { value: 'name', label: 'Name    ' },
  { value: 'quota', label: 'Quota    ' },
];

type list = {
  data: any[];
  onChangeSort: (data: any[]) => void;
  reset: () => void;
};

type ItemType = {
  id: string;
};

const ListDevice = ({ data, onChangeSort, reset }: list) => {
  const initialState = {
    sortOpen: false,
    selectedSort: '',
    refresh: false,
    selectedInfo: '',
  };

  const [state, setstate] = useState(initialState);

  //Sort
  const handleSortChange = (sortOption: string) => {
    let sortedModems: any = [...data];

    switch (sortOption) {
      case 'quota':
        sortedModems.sort(
          (a: any, b: any) =>
            b.quota.maxUsage -
            b.quota.currentUsage -
            (a.quota.maxUsage - a.quota.currentUsage),
        );
        break;
      case 'name':
        sortedModems.sort((a: any, b: any) => a.name.localeCompare(b.name));
        break;
      default:
        reset();
        break;
    }

    setstate(p => ({
      ...p,
      selectedSort: sortOption,
      sortOpen: !state.sortOpen,
    }));

    if (sortOption !== '') {
      onChangeSort(sortedModems);
      setstate(p => ({
        ...p,
        selectedInfo: '',
      }));
    }
  };

  //Select More Info
  const onChangeSelect = (data: any) => {
    setstate(p => ({
      ...p,
      selectedInfo: state.selectedInfo == data ? '' : data,
    }));
  };

  const listHeaderComponent = () => {
    const filterSelect = () => {
      return filter.find(item => item.value === state.selectedSort);
    };

    return (
      <View style={styles.header}>
        <View style={styles.contentHeader}>
          <Image
            source={require('../assets/sort.png')}
            style={styles.sortIcon}
          />
          <Text style={{ color: '#3A4144' }}>Sort By:</Text>
          <View
            style={{
              flex: 1,
              alignItems: 'flex-end',
              paddingHorizontal: 5,
            }}>
            <TouchableOpacity
              onPress={() =>
                setstate(p => ({
                  ...p,
                  sortOpen: !state.sortOpen,
                }))
              }
              style={styles.sort}>
              <Text style={styles.textLabel}>{filterSelect()?.label}</Text>
              <Image
                resizeMode="contain"
                style={{ width: '10%', height: '100%' }}
                source={require('../assets/arrow_bottom.png')}
              />
            </TouchableOpacity>
          </View>
        </View>
      </View>
    );
  };

  const convertToGB = (number: number, unit: string) => {
    let convert: number = number;
    switch (unit) {
      case 'MB':
        convert = convert / 1000;
        break;
      case 'KB':
        convert = convert / 1000000;
        break;
      default:
        break;
    }

    return Math.round(convert);
  };

  const renderItem: ListRenderItem<ItemType> = ({ item, index }: any) => {
    const percentage = Math.round(
      (item.quota.currentUsage / item.quota.maxUsage) * 100,
    );

    return (
      <View style={styles.itemContainer}>
        <View style={styles.item}>
          <View style={{ flex: 2, justifyContent: 'center' }}>
            <Text>name</Text>
            <Text style={styles.textTitle}>{item.name}</Text>
          </View>
          <View style={{ flex: 1, justifyContent: 'center' }}>
            <Text>Quota</Text>
            <View style={styles.progressContain}>
              <View
                style={{
                  backgroundColor:
                    percentage <= 25
                      ? '#DB2424'
                      : percentage > 25 && percentage <= 75
                        ? '#FFCE20'
                        : '#A9CF3B',
                  borderRadius: 32,
                  width: `${percentage}%`,
                  height: 15,
                }}
              />
            </View>
          </View>
          <View style={{ flex: 0, margin: 10, justifyContent: 'center' }}>
            <TouchableOpacity onPress={() => onChangeSelect(item.id)}>
              <Image
                resizeMethod="resize"
                resizeMode="contain"
                style={{ width: 15 }}
                source={
                  state.selectedInfo == item.id
                    ? require('../assets/minus.png')
                    : require('../assets/plus.png')
                }
              />
            </TouchableOpacity>
          </View>
        </View>
        {state.selectedInfo == item.id && (
          <View style={styles.subContain}>
            <View style={styles.subItem}>
              <Text style={{ flex: 1 }}>Usage</Text>
              <Text>
                {convertToGB(item.quota.currentUsage, item.quota.unit)}GB
              </Text>
              <Text style={{ fontWeight: 'bold' }}>
                {` / `}
                {convertToGB(item.quota.maxUsage, item.quota.unit)}GB
              </Text>
            </View>
            <View style={styles.subItem}>
              <Text>Validity</Text>
              <Text>{`${item.validity.value} ${item.validity.unit}`}</Text>
            </View>
          </View>
        )}
      </View>
    );
  };

  const onRefresh = () => {
    setstate(p => ({ ...p, refresh: true }));
    reset();
    setTimeout(() => {
      setstate(p => ({
        ...p,
        refresh: false,
        selectedSort: '',
        sortOpen: false,
      }));
    }, 1000);
  };
  return (
    <View style={{ flex: 1 }}>
      <FlatList
        onRefresh={onRefresh}
        testID='list-device'
        refreshing={state.refresh}
        ListHeaderComponent={listHeaderComponent}
        data={data}
        renderItem={renderItem}
        keyExtractor={(item, index) => index.toString()}
      />

      <SortDevice
        data={filter}
        open={state.sortOpen}
        onClose={() =>
          setstate(p => ({
            ...p,
            sortOpen: false,
          }))
        }
        onChangeSort={handleSortChange}
        sortSelect={state.selectedSort}
      />
    </View>
  );
};
export default ListDevice;
const styles = StyleSheet.create({
  contentHeader: {
    height: height * 0.1,
    width: '90%',
    alignItems: 'center',
    flexDirection: 'row',
    borderBottomWidth: 1,
    borderColor: '#F0F0F0',
  },
  header: {
    alignItems: 'center',
    paddingVertical: 10,
  },
  sortIcon: {
    resizeMode: 'contain',
    width: '4%',
    marginHorizontal: 5,
  },
  labelFilter: {
    borderBottomWidth: 2,
    borderBottomColor: '#E5E5E5',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  sort: {
    minWidth: '40%',
    backgroundColor: '#F0F0F0',
    paddingVertical: 10,
    paddingHorizontal: 20,
    flexDirection: 'row',
    borderRadius: 10,
    justifyContent: 'space-between',
  },
  modalContainer: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  modalContent: {
    backgroundColor: 'white',
    margin: -5,
    borderTopLeftRadius: 16,
    borderTopRightRadius: 16,
    padding: 20,
    width: '100%',
    alignItems: 'center',
  },
  item: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    height: height / 14,
  },
  itemContainer: {
    marginVertical: 10,
    shadowColor: '#000',
    marginHorizontal: 20,

    backgroundColor: 'white',
    borderRadius: 10,
    padding: 10,
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 3,
  },
  textSort: { fontWeight: '400', color: '#262627', marginVertical: 10 },
  textTitle: { fontWeight: '900', color: '#3A4144' },
  progressContain: {
    borderRadius: 32,
    width: '90%',
    backgroundColor: '#EDEDED',
  },
  subItem: {
    flexDirection: 'row',
    flex: 1,
    marginVertical: 10,
    justifyContent: 'space-between',
  },
  subContain: {
    borderTopWidth: 1,
    borderTopColor: '#EBEBEB',
    marginVertical: 5,
  },
  textLabel: { fontWeight: 'bold', alignSelf: 'center', color: 'black' },
});
