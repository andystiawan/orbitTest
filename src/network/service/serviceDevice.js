import {get} from '../helper';
import json2 from '../json/orbit-fe-test-2.json';

export const deviceList = async () => {
  return await get('https://s3-preprod.myorbit.id/files/orbit-fe-test-2.json');
};
