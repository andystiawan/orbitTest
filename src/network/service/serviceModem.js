import {get} from '../helper';
import json1 from '../json/orbit-fe-test-1.json';

export const modemList = async () => {
  return await get('https://s3-preprod.myorbit.id/files/orbit-fe-test-1.json');
};
