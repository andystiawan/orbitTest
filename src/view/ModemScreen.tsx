import React, { useEffect, useState } from 'react';
import { View } from 'react-native';
import { useIsFocused } from '@react-navigation/native';

import { getModemList } from '../network/service/serviceModem';

import CheckoutButton from '../component/CheckoutButton';
import Headers from '../component/Headers';
import ListModem from '../component/ListModem';
import { LoadModem } from '../component/Loading';
import { CheckoutSuccessModal } from '../component/PopUp';

const ModemScreen = () => {

   const initialState = {
      isLoading: false,
      isCheckout: false,
      modemListData: [],
      totalCheckout: 0,

   };

   const [state, setState] = useState(initialState);

   const {
      isLoading,
      isCheckout,
      modemListData,
      totalCheckout,
   } = state

   const isFocused = useIsFocused();

   const fetchData = async () => {
      setState({ ...initialState, isLoading: true });
      const res = await getModemList().then(async res => {
         const result = await JSON.parse(res).data;
         return result.map((item: any) => {
            return {
               id: item?.id,
               stock: item?.Stock,
               name: item?.Name,
               price: item?.Price,
            };
         });
      });

      setTimeout(() => {
         setState(p => ({ ...p, isLoading: false, modemListData: res }));
      }, 3000);
   };

   useEffect(() => {
      isFocused && fetchData();
   }, [isFocused]);


   const calculated = (value: any) => {
      let data: any = [...modemListData];

      const { stock = 0, price = 0, quantity } = value?.modem

      const typePlus = value.type === '+'
      const typeMinus = value.type === '-'

      const result = (qty: number) => ({
         ...value.modem,
         quantity: qty,
         subtotal: price * qty,
      })


      if (typePlus) {
         const qty = quantity ? quantity + 1 : 1;

         data[value.index] = result(qty);
      } else if (typeMinus) {
         const qty = quantity ? quantity - 1 : 0;

         data[value.index] = result(qty);
      } else {
         const qty = Number(value?.value || 0);
         const cek = qty >= stock ? stock : qty;

         data[value.index] = result(cek);
      }

      const findSubtotal = data.filter((x: any) => x?.subtotal);
      const cektotal = calculateTotal(findSubtotal);

      setState({
         ...state,
         modemListData: data,
         totalCheckout: cektotal,
         isCheckout: false,
      });
   };

   const onChangeState = (name: string, value: any) => {
      setState({ ...state, [name]: value })
   }

   const calculateTotal = (items: any) => {
      let total = 0;

      for (const item of items) {
         total += item.subtotal;
      }
      return total;
   };

   const onCheckout = () => {
      onChangeState('isCheckout', true)
   };

   const closeCheckout = () => {
      onChangeState('isCheckout', false)
   };

   return (
      <View style={{ flex: 1 }}>

         <Headers title="Modem List" name="modemListData" totalCheckoutmary={modemListData?.length} />

         {!isLoading ? (
            <View style={{ flex: 1 }}>

               <ListModem
                  checkout={value => calculated(value)}
                  dataFilter={(data: any[]) => setState((p: any) => ({ ...p, modemListData: data }))}
                  data={modemListData}
                  reset={() => fetchData()}
               />

               <CheckoutButton
                  total={totalCheckout}
                  reset={() => fetchData()}
                  onCheckout={() => onCheckout()}
               />

            </View>
         ) : (
            <LoadModem />
         )}

         <CheckoutSuccessModal
            open={isCheckout}
            onClose={closeCheckout}
            total={totalCheckout}
            modemTotal={modemListData.filter((x: any) => x?.subtotal).length}
            onCheckoutButton={closeCheckout}
         />
      </View>
   );
};

export default ModemScreen;
