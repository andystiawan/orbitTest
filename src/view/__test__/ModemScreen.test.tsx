import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { useIsFocused as mockUseIsFocused } from '@react-navigation/native';

import { getModemList } from '../../network/service/serviceModem';

import ModemScreen from '../ModemScreen';
import ListDevice from '../../component/ListDevice';

jest.mock('../../network/service/serviceModem', () => ({
    getModemList: jest.fn()
}));

// Mock useIsFocused
jest.mock('@react-navigation/native', () => ({
    ...jest.requireActual('@react-navigation/native'),
    useIsFocused: jest.fn(),
}));

describe('ModemScreen', () => {

    const data = [
        {
            id: 11,
            name: "Mel's Orbit",
            quota: {
                currentUsage: 30000,
                maxUsage: 80000,
                unit: "MB"
            },
            validity: {
                value: 30,
                unit: "Days"
            }
        },
        {
            id: 23,
            name: "Kitchen Modem",
            quota: {
                currentUsage: 55000,
                maxUsage: 60000,
                unit: "MB"
            },
            validity: {
                value: 7,
                unit: "Days"
            }
        },
        {
            id: 27,
            name: "Donski Pro",
            quota: {
                currentUsage: 25000,
                maxUsage: 45000,
                unit: "MB"
            },
            validity: {
                value: 30,
                unit: "Days"
            }
        },
    ]

    beforeEach(() => {
        (mockUseIsFocused as jest.Mock).mockReturnValue(true);
    });

    it('renders correctly', () => {
        const { toJSON } = render(<ModemScreen />);
        expect(toJSON()).toMatchSnapshot();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders ModemScreen correctly', async () => {
        (getModemList as jest.Mock).mockResolvedValueOnce(JSON.stringify({ data: [] }));

        render(<ModemScreen />);

        expect(getModemList).toHaveBeenCalledTimes(1);
        expect(mockUseIsFocused).toHaveBeenCalledTimes(2);
        expect(mockUseIsFocused).toHaveBeenCalledWith();
    });


    it('should render the ListDevice component with the provided data', () => {
        const onChangeSort = jest.fn()
        const reset = jest.fn()

        const { getByTestId } = render(<ListDevice data={data} onChangeSort={onChangeSort} reset={reset} />)

        const listDevice = getByTestId('list-device')

        expect(listDevice).toBeTruthy();
    });


});