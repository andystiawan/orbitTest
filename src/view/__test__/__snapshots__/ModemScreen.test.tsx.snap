// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModemScreen renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#D81A3C",
          "borderBottomLeftRadius": 16,
          "borderBottomRightRadius": 16,
          "flexDirection": "row",
          "height": 133.4,
          "width": 750,
        },
        false,
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../src/assets/mobile_white.png",
        }
      }
      style={
        [
          {
            "flex": 0,
            "height": "50%",
            "objectFit": "scale-down",
            "resizeMode": "center",
            "width": "20%",
          },
          false,
        ]
      }
    />
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "white",
              "fontWeight": "900",
            },
            false,
          ]
        }
      >
        Modem List
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "top": "-5%",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../src/assets/mobile_black.png",
        }
      }
      style={
        {
          "height": "20%",
          "margin": 15,
          "resizeMode": "contain",
          "width": "20%",
        }
      }
    />
    <Text
      style={
        {
          "fontWeight": "900",
        }
      }
    >
      Loading Modem Data
    </Text>
    <Text
      style={
        {
          "fontWeight": "500",
        }
      }
    >
      Please wait...
    </Text>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
